bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/UsbGatewayDiscoveryService.java

```diff
@@ -0,0 +1,246 @@
+import org.openwebnet4j.OpenDeviceType;
+import org.openwebnet4j.USBGateway;
+import org.openwebnet4j.communication.OWNException;
+import org.openwebnet4j.message.BaseOpenMessage;
+import org.openwebnet4j.message.OpenMessage;
+import org.openwebnet4j.message.Where;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link UsbGatewayDiscoveryService} extends {@link AbstractDiscoveryService} to detect ZigBee USB gateways
+ * connected via serial port. The service will iterate over the available serial ports and open each one to test if a
+ * OpenWebNet ZigBee USB gateway is connected. On successful connection, a new DiscoveryResult is created.
+ *
+ * @author Massimo Valla - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = DiscoveryService.class, configurationPid = "discovery.openwebnet")
+
+public class UsbGatewayDiscoveryService extends AbstractDiscoveryService implements GatewayListener {
+
+    private final Logger logger = LoggerFactory.getLogger(UsbGatewayDiscoveryService.class);
+
+    private static final int DISCOVERY_TIMEOUT_SECONDS = 30;
+    private static final int PORT_CHECK_TIMEOUT_MSEC = 1500;
+
+    private CountDownLatch portCheckLatch = new CountDownLatch(1);
+    private @Nullable ScheduledFuture<?> connectTimeout;
+
+    private @NonNullByDefault({}) SerialPortManager serialPortManager;
+    private @Nullable USBGateway zbGateway;
+
+    private String currentScannedPortName = "";
+
+    /**
+     * Keeps a boolean during time discovery process in progress.
+     */
+    private boolean scanning;
+
+    /**
+     * Constructs a new UsbGatewayDiscoveryService with the specified ZigBee USB Bridge ThingTypeUID
+     */
+    public UsbGatewayDiscoveryService() {
+        super(new HashSet<>(Arrays.asList(OpenWebNetBindingConstants.THING_TYPE_ZB_GATEWAY)), DISCOVERY_TIMEOUT_SECONDS,
+                false);
+    }
+
+    /**
+     * Starts a new discovery scan. All available Serial Ports are scanned.
+     */
+    @Override
+    protected void startScan() {
+        logger.debug("Started OpenWebNet ZigBee USB Gateway discovery scan");
+        removeOlderResults(getTimestampOfLastScan());
+        scanning = true;
+        Stream<SerialPortIdentifier> portEnum = serialPortManager.getIdentifiers();
+        // Check each available serial port
+        portEnum.forEach(portIdentifier -> {
+            if (scanning) {
+                currentScannedPortName = portIdentifier.getName();
+                logger.debug("[{}] == checking serial port", currentScannedPortName);
+                if (portIdentifier.isCurrentlyOwned()) {
+                    logger.debug("[{}] serial port is owned by: {}", currentScannedPortName,
+                            portIdentifier.getCurrentOwner());
+                    if ("NRSerialPort".equals(portIdentifier.getCurrentOwner())) {
```

Comment:
This check relys on a library deep inside OH. If the library is changed, this check will break. Maybe you can express this a bit more abstract to avoid the `if`.

Reply:
I do not see an easy way to check if the serial port being scanned is owned by this binding itself. I have no reference here in this class to the possible configured bridge class that is using the same serial port. Moreover, this is exactly what is being done here https://github.com/openhab/openhab-addons/blob/0fdf1e34c31aec84dc0b78d355d508e219e7e793/bundles/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java#L112
@Hilbrand suggested in the past to use DSMR binding as an example for serial USB device discovery.
Any suggestion to follow?

Reply:
This should be ok. However it should compare to the name set by this binding when opening the serial port ( I didn't check what this binding uses, but give it's a string here and not a constant might mean it's some kind of default name that is checked here or it sets this generic name). This name should relate to the binding (and preferable be put into a constant that is also used when opening the port). See how the DSMR binding does this)

Reply:
There is currently no support in the underlying lib (openwebnet4j) for setting a custom port owner so the default one is used NRSerialPort : it would need to add it as a constructor param in 2 classes. Moreover it’s not properly supported by nrjavaserial as isCurrentlyOwned fails most of the time (reporting false when it should report true).
 If you think this is really important ok, I will change all needed classes (2h +testing)... it was just a courtesy message in the rare case the lib hangs and the port is kept occupied and the binding tries to connect at it

Reply:
I'd suggest to simply remove the courtesy message or express it a bit more abstract so that you don't need to check for the magic string.

Reply:
ok, I will remove the message then, as nrjavaserial does not support setting the owner string, it just sets `NRSerialPort ` by default.